rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,78 @@
+### DCP PR:
+
+# Metadata integration testing
+
+## Summary
+
+This document details an integration test design for metadata schema changes in the Human Cell Atlas (HCA) Data Coordination Platform (DCP).
+
+## Authors
+
+* [Andrey Kislyuk](mailto:akislyuk@chanzuckerberg.com)
+* [Sam Pierson](mailto:spierson@chanzuckerberg.com)
+* [Matt Weiden](mailto:mweiden@chanzuckerberg.com)
+* [Mark Diekhans](mailto:markd@ucsc.edu)
+* [Hannes Schmidt](mailto:hannes@ucsc.edu)
+
+## Shepherd
+
+## Motivation
+
+Currently breakages to DCP systems caused by metadata schema changes are detected at runtime upon data (bundle) upload. This forces data wranglers, DCP operators, and data consumers to react to unpredictable runtime failures rather than detecting problems before the code is merged and deployed.
+
+To alleviate this issue, the DCP development team can use a metadata schema integration test to test whether it is safe to release data downstream of the Data Storage Service (DSS), past which point applications start to depend on the schema. The benefits of this approach are that it will:
+
+1. Make the development environment more stable
+1. Make updates to software and schema smaller and more frequent
+1. Make failures due to metadata schema changes more readily observable
+1. Increase availability of DCP systems downstream of DSS
+
+## Detailed Design
+
+In this design we will use the following terms:
+ 
+ * "schema integration test system" - an automated CI/CD pipeline used to check that downstream systems can handle bundles with a new schema
+ * "new schema label" - anything that indicates that the schema has changes that have not yet been tested against downstream systems: potentially a JSON field, git tag, git branch, or comparison to tested version list
+
+
+Design
+
+1. Bundles with new metadata schema changes are uploaded to the production environment with a new schema label
+1. Observing the new schema label, the DSS stores the new bundles and makes them available with the `GET /v1/bundles` endpoint with the bundle `uuid` or `(uuid, version)`, but does not index nor does it release subscription notifications for them.
+1. Via a schema integration test system, a sample of the uploaded bundles are copied from the production DSS to the integration DSS testing environment.
```

Comment:
Good point. What strategies for selecting the sample would you suggest?

Reply:
I don't think there's a simple strategy for working out what you need to test without just testing all the data. You could select a sample of data for every unique graph shape in the experiment (we're working on tooling to identify different shapes). You might also want to select samples that use specific fields that we know have been either added or changed in the schema, but this would require some level of schema introspection.

Reply:
Sounds like a great opportunity to apply the tooling you're building!